components/support/sql/src/conn_ext.rs

```diff
@@ -213,3 +217,72 @@ impl<'conn> ConnExt for UncheckedTransaction<'conn> {
         &*self
     }
 }
+
+/// This transaction is suitable for when a transaction is being used only
+/// for performance when updating a larger number of rows (ie, when the
+/// transaction is not necessary for data integrity). You can specify a
+/// duration for the maximum amount of time the transaction should be held.
+/// You should regularly call .maybe_commit() as part of your processing, and
+/// if the current transaction has been open for greater than the specified
+/// duration the transaction will be committed and another one started.
+/// You should always call .commit() at the end. Note that if you call
+/// .rollback() at the end, only the current transaction will be rolled back.
+pub struct TransactionForPerf<'conn> {
+    tx: UncheckedTransaction<'conn>,
+    behavior: TransactionBehavior,
+    commit_after: Duration,
+}
+
+impl<'conn> TransactionForPerf<'conn> {
+    /// Begin a new transaction which may be split into multiple transactions
+    /// for performance reasons. Cannot be nested, but this is not
+    /// enforced; use a rusqlite `savepoint` for nested transactions.
+    pub fn new(
+        conn: &'conn Connection,
+        behavior: TransactionBehavior,
+        commit_after: Duration,
+    ) -> SqlResult<Self> {
+        Ok(Self {
+            tx: UncheckedTransaction::new(conn, behavior)?,
+            behavior,
+            commit_after,
+        })
+    }
+
+    /// Checks to see if we have held a transaction for longer than the
+    /// requested time, and if so, commits the current transaction and opens
+    /// another.
+    pub fn maybe_commit(&mut self) -> SqlResult<()> {
+        if self.tx.started_at.elapsed() >= self.commit_after {
+            self.tx.conn.execute_batch("COMMIT")?;
+            log::debug!("Transaction commited after taking allocated time");
+            self.tx = UncheckedTransaction::new(self.tx.conn, self.behavior)?;
```

Comment:
Hmm, I feel uneasy about committing the transaction out from under the `UncheckedTransaction` here.

Why not use a `Cell<Option<UncheckedTransaction>>` to store `tx`, so that we can consume `self.tx` here (`self.tx.take().unwrap().commit()`, but with better error handling), then `self.tx.set(Some(UncheckedTransaction::new(...)?))` to store the new transaction back in the cell. That way, if we fail to start a new transaction because another writer beat us to it, we won't hold on to the (now-stale) `UncheckedTransaction` in `tx`. And we could then return errors if you try to `commit` or `rollback` when we don't have a `tx`, instead of silently committing or rolling back whatever's in progress.

Do you think that might work?

Reply:
You could do this without the Cell<Option> with `std::mem::swap`. E.g.
```rust
let mut tx = UncheckedTransaction::new(...);
std::mem::swap(&mut self.tx, &mut tx);
tx.commit()?;
```